### PR TITLE
Parallel team functor

### PR DESCRIPTION
**Pull request by Rombur**
This adds `parallel_for` for Team. A couple of things to note:
 - to break circular dependency I have moved code from `core/src/Kokkos_HIP.hpp` to `core/src/Kokkos_HIP_Space.hpp`. Now `core/src/Kokkos_HIP.hpp` only includes other headers, it doesn't have any code in it.
 - I have not implemented the functions that compute the optimal block size. The algorithms will be different that for CUDA because the properties we get are different. Also for now, what hip reports is bugged.
 - Due to a current bug in hipcc, we cannot use global variables for the lock without RDC but linking with RDC is extremely slow. I wrote a workaround that allocate/deallocate memory when the kernel is launched but everything is implemented to use global variable. So when AMD fixes the bug we should be able to just remove the workaround and uncomment one line and be done.

### CODE REVIEW

**Comment by Rombur** [comment]
I forgot to turn on the warnings when developing so I need to do something about that. I am not sure why the tester is not happy, it works locally. I am looking into it.

File: core/src/HIP/Kokkos_HIP_Locks.hpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    if (::Kokkos::Impl::lock_array_copied == 0) {                           \
+      HIP_SAFE_CALL(hipMemcpyToSymbol(HIP_SYMBOL(g_device_hip_lock_arrays), \
+                                      &g_host_hip_lock_arrays,              \
+                                      sizeof(HIPLockArrays)));              \
+    }                                                                       \
+    lock_array_copied = 1;                                                  \
+  }
+
+#ifdef KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE
+#define KOKKOS_ENSUREG_HIP_LOCK_ARRAYS_ON_DEVICE()
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Locks.hpp]: ENSURE[G]

File: core/src/HIP/Kokkos_HIP_Locks.hpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+namespace Kokkos {
+namespace Impl {
+
+/// \brief Acquire a lock for the address
+///
+/// This function tries to acquire the lock for the hash value derived
+/// from the provided ptr. If the lock is successfully acquired the
+/// function returns true. Otherwise it returns false.
+__device__ inline bool lock_address_hip_space(void* ptr) {
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Locks.hpp]: AFAICT, this function is not used in this pull request at all. Apart from the definition here, I only find one more occurrence in `Kokkos_HIP_Atomic.hpp` in commented code.
Am I missing something here or should this be in a separate pull request?

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp [hunk 3]
```diff
@@ -117,6 +133,203 @@ struct HIPGetMaxBlockSize<DriverType, Kokkos::LaunchBounds<>, true> {
...
+                           hipFuncAttributes const &attr,
+                           FunctorType const & /*f*/,
+                           size_t const /*vector_length*/,
+                           size_t const /*shmem_block*/,
+                           size_t const /*shmem_thread*/) {
+  // FIXME_HIP find a better algorithm. Be aware that
+  // maxThreadsPerMultiProcessor, regsPerBlock, and l2CacheSize are bugged and
+  // always return zero
+  // https://github.com/ROCm-Developer-Tools/HIP/blob/6c5fa32815650cc20a4f783d09b013610348a4d5/include/hip/hcc_detail/hip_runtime_api.h#L438-L440
+  // and we don't have access to the same information than we do for CUDA
```
[the remark this answers is not in the document]
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp]: I didn't report it since they already have a comment about the bugs.

File: core/src/HIP/Kokkos_HIP_Locks.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/HIP/Kokkos_HIP_Locks.hpp]: I still think that it should be in a separate pull request then.
